src/compiler/checker.ts

```diff
@@ -25299,14 +25302,33 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
         return links.type;
     }
 
-    function inferReverseMappedType(sourceType: Type, target: MappedType, constraint: IndexType): Type {
+    function inferReverseMappedTypeWorker(sourceType: Type, target: MappedType, constraint: IndexType): Type {
         const typeParameter = getIndexedAccessType(constraint.type, getTypeParameterFromMappedType(target)) as TypeParameter;
         const templateType = getTemplateTypeFromMappedType(target);
         const inference = createInferenceInfo(typeParameter);
         inferTypes([inference], sourceType, templateType);
         return getTypeFromInference(inference) || unknownType;
     }
 
+    function inferReverseMappedType(source: Type, target: MappedType, constraint: IndexType): Type {

```

Comment:
Should we cache the results inside this, so we guarantee deeply nested results get cached? Of the callers of this function, only `getTypeOfReverseMappedSymbol` currently caches - the other uses don't (directly) cache the result, so could, in theory, witness different results for the same set of input types at different points. You can probably only easily trigger such a behavior with a reverse mapped type over an array...? Maybe?

Reply:
I was adding the cache to `inferReverseMappedType` and looking at `inferTypeForHomomorphicMappedType`. I'm now wondering: could I do away with the recursion detection in `inferTypeForHomomorphicMappedType` (that uses `homomorphicMappedTypeInferenceStack`)?

Reply:
I'm pretty sure that we can, however this once again begs the question of what to return when we find a circularity. `inferTypeForHomomorphicMappedType` returns `undefined` in this case. I'll have to think some more about this, but I'm planning on removing that circularity detection code in `inferTypeForHomomorphicMappedType`.